### ISSUE TITLE
valhalla_build_admins strict mode suggestion

### ISSUE THREAD

**Issue by ianthetechie** [valhalla/valhalla#3721]
Half-question, half-issue here. I noticed that with the name change to Türkiye, `valhalla_build_admins` just put a nice little error in the log and moved on with life. This isn't the end of the world, but looking deeper into `BuildAdminFromPBF`, I noticed that the function doesn't return any value indicating success/failure, and as a result, the CLI utility pretty much always sets a successful exit status.

I personally find this pretty counterintuitive, since there appear to be at least a dozen reasons why the process could fail. I would be open to contributing code to make this a bit more communicative. At a minimum, I think the `valhalla_build_admins` CLI utility should return a failure exit status in most obvious error cases (most of the failures possible), and `BuildAdminFromPBF` should return a boolean indicating whether it was successful.

The majority of the early returns are clear failures. This specific example with a country name change is an interesting one though, since as the docs say, this could arise normally when dealing with a partial extract. So perhaps there are two modes: a normal mode and a more "pedantic" one? Not at all sold on my choice of term, so input very much welcome. The normal mode would treat any of the existing early returns as a failure, communicate this via a return value, and `valhalla_build_admins` would exit with a failed status. In pedantic mode, places such as this which currently `LOG_ERROR` but don't bail out now would also bail and fail.

Thoughts?

**Comment by kevinkreiser** [comment]
@ianthetechie i've currently re-written the admin import and added a ton more logging to it when things go wrong and where: #3683 so they can be looked into at a data level

as for your specific issue, are you using a planet or extract that was created before the name change?

regarding exit status, its deliberately liberal, letting almost anything pass. a couple of examples on when and why this happens:

1. you have an osm extract there could be partial relations in there so we cant reconstruct the whole polygon, should we fail the whole build or let you get out the polygons which we could recover?
2. someone screwed up the relation in osm, either they didnt connect the thing, they did something geometrically degenerate or they just put the wrong osm ids in for member ways or member nodes. again, should we warn about this one screwed up admin and give you back as much as we can or should we fail the whole thing?

the basic crux is we cant make any assumptions about the completeness or correctness of the data. instead we tell you when stuff goes wrong (latest pr tells even more) and we give you back as much as we can from the data you provided. if instead we encountered problems and bailed as soon as we hit them (pedantic/strict mode), you pretty much would never get an admin db built. in my experience ive never seen an exit code used to communicate partial success but is that what you are suggesting here?

also i think you have pedantic and normal switched. pedantic is more of an above and beyond normal, its like a strict mode. but in any case i am almost certain that having a strict mode will be useless because, like i said, you wont find a planet, unless you made it from HERE or TomTom data for example, that doesnt have some kind of problems with it.

am i making sense?

**Comment by ianthetechie** [comment]
Hey Kevin, thanks for the response! Definitely appreciate the extra logging in #3683!

I am using the latest weekly planet (220815) dump, but was using previous *release* code. It of course works fine against `master` given the changes merged a few days ago, but I am really just looking to see if there's a better way to catch errors without having to grep logs :joy: The issue was just related to the name change not matching the hard-coded constants in the [header](https://github.com/valhalla/valhalla/blob/9c06fece397e04e4971a5a759596ea1315c553ac/valhalla/mjolnir/adminconstants.h#L159) until a few days ago when the fix was merged.

Sorry if my suggestion wasn't clear; I'll try to be a bit more explicit.

1. Add a strict mode (I like that term better; thanks for the suggestion) CLI flag (defaults to false)
2. Pass this strict flag to `BuildAdminFromPBF`; when true, return early with a `false` value if anything is not clean, incl. being unable to add info due to name changes, which do not currently bail early. Normal (non-strict) mode would not return early in the case of name changes, just as it does today.
3. In strict mode, the exit code would indicate any failure.

> if instead we encountered problems and bailed as soon as we hit them (pedantic/strict mode), you pretty much would never get an admin db built.

I'll defer to you on this. I haven't actually paid too close attention to these for a while and it clearly hasn't mattered much, but I sometimes get carried away in the pursuit of perfection :joy: FWIW, there are no error logs generating admins from the latest planet dump using `master`. I think an optional strict mode as a canary to catch obvious things that can be fixed easily would be beneficial, but that's just my 2 cents.

TL;DR, I'm not suggesting an exit code to indicate partial success; I'm only suggesting using the exit code in strict mode to indicate that a totally clean admin DB couldn't be generated.

**Comment by nilsnolde** [comment]
I do like the suggestion. I hardly ever see the admin log of a planet build, and got too accustomed to the many errors of processing extracts. It the planet builds cleanly I’d like to enable strict mode on the public instance. Even a mapping error would be very valuable info for OSM as a whole. I don’t think there’s many services processing the high level admin information on the same rotation as us (~36 hours). 

The Turkey change was told to me by a customer and was already days or weeks old. But I don’t check the public instance unless it’s failing. And IMHO this OSM data QA/QC could be an important aspect for the public fossgis instance, and also part of an argument (albeit small) to keep running it (we’ll soon have to ask for extension).

**Comment by kevinkreiser** [comment]
> FWIW, there are no error logs generating admins from the latest planet dump using master

thats because the old code didnt log much :D for example, when there were inners which didn't find any outers to nest within they were promoted to outers silently. if a multipolygon came out and couldnt be made valid, it was silently skipped in the output. for polygon rings, if they had self intersections its unclear that finding their inners would work properly and there is no attempt made to correct any polygon rings at all. again if this causes an inner to miss an outer its not logged the inner just becomes an outer. i could go back and review the code more but these are at least some serious issues.

with the latest pr you will see there are lots of places where this stuff is at least now logged so that someone can take action to fix the data.

so all that said, we can make a strict mode, but i'm skeptical it will be useful because i am fairly confident it will often fail. anyway defaulting to not strict mode is fine then. to me t hough id rather always run in unstrict mode and see all of the errors that need to be fixed than run in strict mode, have it fail on the first error, go fix, apply my changes to my planet, ru n again, see that it errors on something else and do it again and again. im just struggling to see the use case.

if we really want to find and fix problems we should do what we did for the map roulette integration. add an option to output errors to (maybe a geojson) file which then can be investigated after the program runs. its solves the problem of you having to grep stdout but you would still have to see if there is anything in the optional output file.

**Comment by ianthetechie** [comment]
Gotcha. I think I understand what you're saying now. I'll revisit this once your branch is merged

**Comment by nilsnolde** [comment]
Looking at @kevinkreiser's [recent comment](https://github.com/valhalla/valhalla/pull/3683#issuecomment-1226705415) on the geos PR, I withdraw my last comment and fully agree there's no point in a (general) strict mode. Catching name changes is a more manageable error we could fail on in a strict mode. WDYT?

Name changes do happen irregularly, in the past 6 months twice, Türkiye and Saint-Martin (in #3619). The last one we didn't catch for multiple months.

**Comment by ianthetechie** [comment]
Yeah, I'm coming to the same conclusion as well. This incidentally makes things like [Daylight Distribution](https://daylightmap.org/) rather appealing over raw OSM extracts. Closing this issue.
